analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantNullCheckCodeFixProvider.cs

```diff
@@ -24,63 +24,89 @@
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using SonarAnalyzer.Extensions;
 using SonarAnalyzer.Helpers;
+using StyleCop.Analyzers.Lightup;
 
 namespace SonarAnalyzer.Rules.CSharp
 {
     [ExportCodeFixProvider(LanguageNames.CSharp)]
     public sealed class RedundantNullCheckCodeFixProvider : SonarCodeFixProvider
     {
         internal const string Title = "Remove this unnecessary null check";
-        public override ImmutableArray<string> FixableDiagnosticIds
-        {
-            get
-            {
-                return ImmutableArray.Create(RedundantNullCheck.DiagnosticId);
-            }
-        }
 
-        public override FixAllProvider GetFixAllProvider()
-        {
-            return WellKnownFixAllProviders.BatchFixer;
-        }
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(RedundantNullCheck.DiagnosticId);
+
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
 
         protected override Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext context)
         {
             var diagnostic = context.Diagnostics.First();
             var diagnosticSpan = diagnostic.Location.SourceSpan;
+            var diagnosticNode = root.FindNode(diagnosticSpan, getInnermostNodeForTie: true);
 
-            // FIXME: avoid providing a fix for patterns, as it can become messy for non-trivial recursive patterns
-            if (!(root.FindNode(diagnosticSpan, getInnermostNodeForTie: true) is BinaryExpressionSyntax nullCheckNode))
+            if (diagnosticNode is BinaryExpressionSyntax nullCheckNode)
+            {
+                RegisterBinaryExpressionCodeFix(context, root, nullCheckNode);
+            }
+            else if (diagnosticNode is PrefixUnaryExpressionSyntax prefixUnary && prefixUnary.IsKind(SyntaxKind.LogicalNotExpression))
+            {
+                RegisterBinaryExpressionCodeFix(context, root, prefixUnary);
+            }
+            else if (IsPatternExpressionSyntaxWrapper.IsInstance(diagnosticNode))
+            {
+                var isPatternExpression = (IsPatternExpressionSyntaxWrapper)diagnosticNode.RemoveParentheses();
+                if (isPatternExpression.IsNull() || isPatternExpression.IsNotNull())
+                {
+                    RegisterBinaryExpressionCodeFix(context, root, isPatternExpression.SyntaxNode);
+                }
+            }
+            else if (PatternSyntaxWrapper.IsInstance(diagnosticNode))
             {
-                return TaskHelper.CompletedTask;
+                RegisterBinaryPatternCodeFix(context, root, ((PatternSyntaxWrapper)diagnosticNode).SyntaxNode);
             }
 
+            return TaskHelper.CompletedTask;
```

Comment:
Is there a reason why don't we fix 

```
var x = a switch
{
    null or not "a" => "",
//  ^^^^
    _ => ""
};
```

Should be pretty easy to do for nulls in such patterns.

Reply:
good catch!